File: app/src/main/java/com/example/mystepcounterapp/MainActivity.java
package com.example.mystepcounterapp; // <-- ЗАМЕНИ НА ТВОЙ ПАКЕТ

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;

import android.Manifest; // Важно!
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.util.Log;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast; // Для показа сообщений пользователю

import java.util.Locale;

public class MainActivity extends AppCompatActivity {

    private static final String TAG = "MainActivity";

    private Button btnStartService;
    private Button btnStopService;
    private TextView tvSteps;

    // Приемник для получения данных от сервиса
    private BroadcastReceiver stepUpdateReceiver;

    // Код запроса разрешения
    private static final int ACTIVITY_RECOGNITION_PERMISSION_CODE = 100;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main); // Убедись, что layout называется activity_main.xml

        btnStartService = findViewById(R.id.btnStartService);
        btnStopService = findViewById(R.id.btnStopService);
        tvSteps = findViewById(R.id.tvSteps);

        // Устанавливаем обработчики нажатий
        btnStartService.setOnClickListener(v -> tryStartStepService()); // Измененный вызов
        btnStopService.setOnClickListener(v -> stopStepService());

        // Инициализация приемника для сообщений от сервиса
        stepUpdateReceiver = new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {
                if (intent != null && StepService.ACTION_STEP_UPDATE.equals(intent.getAction())) {
                    int steps = intent.getIntExtra(StepService.EXTRA_STEP_COUNT, 0);
                    Log.d(TAG, "Received step update via Broadcast: " + steps);
                    // Обновляем TextView на UI потоке
                    tvSteps.setText(String.format(Locale.getDefault(), "Шагов: %d", steps));
                }
            }
        };
    }

    @Override
    protected void onResume() {
        super.onResume();
        // Регистрируем приемник при возобновлении активности
        LocalBroadcastManager.getInstance(this).registerReceiver(
                stepUpdateReceiver,
                new IntentFilter(StepService.ACTION_STEP_UPDATE) // Используем тот же action, что и в сервисе
        );
        Log.d(TAG, "BroadcastReceiver registered");
    }

    @Override
    protected void onPause() {
        super.onPause();
        // Отписываемся от приемника при сворачивании активности, чтобы не было утечек
        LocalBroadcastManager.getInstance(this).unregisterReceiver(stepUpdateReceiver);
        Log.d(TAG, "BroadcastReceiver unregistered");
    }

    /**
     * Пытается запустить сервис подсчета шагов, проверяя разрешения.
     */
    private void tryStartStepService() {
        // Разрешение ACTIVITY_RECOGNITION требуется начиная с Android 10 (API 29)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            // Проверяем, есть ли уже разрешение
            if (ContextCompat.checkSelfPermission(this, Manifest.permission.ACTIVITY_RECOGNITION)
                    == PackageManager.PERMISSION_GRANTED) {
                // Разрешение есть, запускаем сервис
                Log.d(TAG, "ACTIVITY_RECOGNITION permission already granted.");
                startStepServiceActual();
            } else {
                // Разрешения нет. Проверяем, нужно ли показывать объяснение.
                if (ActivityCompat.shouldShowRequestPermissionRationale(this, Manifest.permission.ACTIVITY_RECOGNITION)) {
                    // Показываем объяснение пользователю (например, в диалоге)
                    // В этом примере просто выводим Toast и запрашиваем снова
                    Log.w(TAG, "Showing rationale for ACTIVITY_RECOGNITION permission.");
                    Toast.makeText(this, "Разрешение на физическую активность нужно для подсчета шагов", Toast.LENGTH_LONG).show();
                    // После показа объяснения, можно снова запросить разрешение
                    requestActivityRecognitionPermission();
                } else {
                    // Объяснение не требуется (первый запрос или пользователь выбрал "Don't ask again")
                    // Просто запрашиваем разрешение
                    Log.d(TAG, "Requesting ACTIVITY_RECOGNITION permission...");
                    requestActivityRecognitionPermission();
                }
            }
        } else {
            // Для версий Android ниже 10 разрешение ACTIVITY_RECOGNITION не требуется, запускаем сразу
            Log.d(TAG, "No need to request ACTIVITY_RECOGNITION permission (API < 29).");
            startStepServiceActual();
        }
    }

    /**
     * Выполняет фактический запрос разрешения ACTIVITY_RECOGNITION.
     */
    private void requestActivityRecognitionPermission() {
        ActivityCompat.requestPermissions(
                this,
                new String[]{Manifest.permission.ACTIVITY_RECOGNITION},
                ACTIVITY_RECOGNITION_PERMISSION_CODE
        );
    }


    /**
     * Метод, который фактически запускает сервис (вызывается после получения разрешений).
     */
    private void startStepServiceActual() {
        // Дополнительно проверим разрешение на уведомления для Android 13+
        // Хотя сервис сам запросит его при startForeground, если нет,
        // лучше запросить заранее для более гладкого UX.
        // Эту проверку можно добавить сюда или оставить как есть.

        Log.d(TAG, "Starting StepService...");
        Intent serviceIntent = new Intent(this, StepService.class);
        // Используем startForegroundService для Android 8.0 (API 26) и выше
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            startForegroundService(serviceIntent);
        } else {
            startService(serviceIntent);
        }
    }

    /**
     * Останавливает сервис подсчета шагов.
     */
    private void stopStepService() {
        Log.d(TAG, "Stopping StepService...");
        Intent serviceIntent = new Intent(this, StepService.class);
        stopService(serviceIntent);
        // Можно сбросить текст шагов здесь при остановке
        // tvSteps.setText(R.string.steps_initial);
    }

    /**
     * Обработка результата запроса разрешений от пользователя.
     */
    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);

        if (requestCode == ACTIVITY_RECOGNITION_PERMISSION_CODE) {
            // Проверяем результат запроса на ACTIVITY_RECOGNITION
            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                // Пользователь предоставил разрешение
                Log.d(TAG, "ACTIVITY_RECOGNITION permission granted by user.");
                // Теперь можно безопасно запускать сервис
                startStepServiceActual();
            } else {
                // Пользователь отклонил разрешение
                Log.w(TAG, "ACTIVITY_RECOGNITION permission denied by user.");
                // Показываем сообщение, что функция недоступна без разрешения
                Toast.makeText(this, "Без разрешения подсчет шагов невозможен", Toast.LENGTH_LONG).show();
                // Можно предложить пользователю перейти в настройки приложения,
                // если он выбрал "Don't ask again".
            }
        }
        // Здесь можно добавить обработку других кодов запросов разрешений, если они будут
    }
}

File: WalkiePaws/app/src/main/java/com/example/walkiepaws/MainActivity.kt
package com.example.walkiepaws

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.widget.Button
import android.widget.EditText
import android.widget.Toast
import androidx.activity.enableEdgeToEdge
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.example.walkiepaws.backend.ApiService
import com.example.walkiepaws.backend.RetrofitClient
import com.example.walkiepaws.backend.model.dto.request.AuthenticationDTO
import com.example.walkiepaws.backend.model.dto.responce.LoginDTO
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class MainActivity : AppCompatActivity() {
    private lateinit var editTextLogin: EditText
    private lateinit var editTextPassword: EditText
    private lateinit var buttonLogin: Button
    private lateinit var buttonRegistration: Button
    private lateinit var apiService: ApiService
    private val TAG: String = "MainActivity"
    private val ACTIVITY_RECOGNITION_PERMISSION_CODE: Int = 100

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        apiService = RetrofitClient.getApiService()
        setContentView(R.layout.activity_main)
        enableEdgeToEdge()
        initViews()
        setupClickListeners()
    }


    private fun initViews() {
        editTextLogin = findViewById(R.id.editTextWriteLogin)
        editTextPassword = findViewById(R.id.editTextWritePassword)
        buttonLogin = findViewById(R.id.buttonLogin)
        buttonRegistration = findViewById(R.id.buttonRegistration)
    }

    private fun setupClickListeners() {
        buttonLogin.setOnClickListener { handleLogin() }
        buttonRegistration.setOnClickListener { navigateToRegistration() }
    }

    private fun handleLogin() {
        val authDTO = AuthenticationDTO(
            editTextLogin.text.toString().trim(),
            editTextPassword.text.toString().trim()
        )

        apiService.auth(authDTO).enqueue(object : Callback<LoginDTO> {
            @RequiresApi(Build.VERSION_CODES.Q)
            override fun onResponse(call: Call<LoginDTO>, response: Response<LoginDTO>) {
                if (response.code() == 200) {
                    val loginResponse = response.body()
                    (applicationContext as App).token = "Bearer " + loginResponse?.token
                    showToast("Вы успешно вошли!")
                    navigateToMainGameScreen()
                } else {
                    showToast("Неверный логин или пароль")
                }
            }

            override fun onFailure(call: Call<LoginDTO>, t: Throwable) {
                t.message?.let { Log.d("DEBUG", it) }
                showToast("Ошибка связи с сервером")
            }
        })
    }

    private fun navigateToRegistration() {
        startActivity(Intent(this, RegistrationActivity::class.java))
    }

    @RequiresApi(Build.VERSION_CODES.Q)
    private fun navigateToMainGameScreen() {
        tryStartStepService()
        DataManager.updateData()
        startActivity(Intent(this, MainGameScreen::class.java))
    }

    private fun showToast(message: String) {
        Toast.makeText(this, message, Toast.LENGTH_SHORT).show()
    }

    @RequiresApi(Build.VERSION_CODES.Q)
    private fun tryStartStepService() {
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.ACTIVITY_RECOGNITION)
            == PackageManager.PERMISSION_GRANTED
        ) {
            Log.d(TAG, "ACTIVITY_RECOGNITION permission already granted.")
            startStepServiceActual()  // Запускаем сервис, если разрешение уже дано
        } else {
            // Запрашиваем разрешение
            if (ActivityCompat.shouldShowRequestPermissionRationale(
                    this,
                    Manifest.permission.ACTIVITY_RECOGNITION
                )
            ) {
                Log.w(TAG, "Showing rationale for ACTIVITY_RECOGNITION permission.")
                Toast.makeText(
                    this,
                    "Разрешение на физическую активность нужно для подсчета шагов",
                    Toast.LENGTH_LONG
                ).show()
                requestActivityRecognitionPermission()
            } else {
                Log.d(TAG, "Requesting ACTIVITY_RECOGNITION permission...")
                requestActivityRecognitionPermission()
            }
        }
    }


    @RequiresApi(Build.VERSION_CODES.Q)
    private fun requestActivityRecognitionPermission() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            // Запрос разрешения
            ActivityCompat.requestPermissions(
                this,
                arrayOf(Manifest.permission.ACTIVITY_RECOGNITION),
                ACTIVITY_RECOGNITION_PERMISSION_CODE
            )
            startStepServiceActual()
        } else {
            onRequestPermissionsResult(
                ACTIVITY_RECOGNITION_PERMISSION_CODE,
                arrayOf(Manifest.permission.ACTIVITY_RECOGNITION),
                intArrayOf(PackageManager.PERMISSION_GRANTED)  // Симулируем, что разрешение предоставлено
            )
            startStepServiceActual()
        }
    }

    private fun startStepServiceActual() {

        Log.d(TAG, "Starting StepService...")
        val serviceIntent = Intent(
            this,
            StepService::class.java
        )
        startForegroundService(serviceIntent)
    }

    private fun stopStepService() {
        Log.d(TAG, "Stopping StepService...")
        val serviceIntent = Intent(
            this,
            StepService::class.java
        )
        stopService(serviceIntent)
    }


    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)

        // Проверка, что мы получаем результат именно для разрешения ACTIVITY_RECOGNITION
        if (requestCode == ACTIVITY_RECOGNITION_PERMISSION_CODE) {
            if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                Log.d(TAG, "ACTIVITY_RECOGNITION permission granted by user.")
                startStepServiceActual()  // Запускаем сервис, если разрешение предоставлено
            } else {
                Log.w(TAG, "ACTIVITY_RECOGNITION permission denied by user.")
                Toast.makeText(
                    this,
                    "Без разрешения подсчет шагов невозможен",
                    Toast.LENGTH_LONG
                ).show()
            }
        }
    }
}